pkg/common/mpool/freelist.go

```diff
@@ -19,21 +19,30 @@ import (
 	"unsafe"
 )
 
+type slot struct {
+	// a triple value is used here to protect
+	// 0: means avaliable
+	// 1: means free
+	// 2: means critical
+	set atomic.Int32
```

Comment:
What kind of ring buffer are we trying to fix here?   I don't think a ring buffer need a atomic flag in slot data structure.

Reply:
This is the easiest way to do it in my opinion, but it is possible to do without this variable, but it would require a change to the original logic.

Reply:
The original logic was that if the rate was fast enough, the ring would come back around and the protection of head, tail was not enough. @fengttt

Reply:
head, tail are not pointers and are not the same thing as the actual slot. So there are problems. Using a pointer directly should circumvent this variable.